### DISCUSSION TITLE
Creating an NPM package using Bun

### DISCUSSION THREAD

**Discussion by pinkboid** [oven-sh/bun#6034] [Q&A]
I want to create a JavaScript library to publish on NPM, but develop it using Bun for runtime testing and dependency management. Is there any official documentation on how to do it? I assume it's exactly the same as you would create an NPM package normally, except you .npmignore all the bun stuff. Any other steps I should take?

**Comment by pinkboid** [comment]
For example, what am I to do about `bun.lockb`? I assume I simply add it it to `.npmignore`, but would like clarification.

**Reply by idleberg** [reply]
Lock files are not part of a package tarball, it is not relevant what you use in your repository. For the same reason the React team (to pick a popular example) can work with a Yarn lock file.

**Comment by bzvestey** [comment]
I am also interested in this. I am looking into converting a set of micro-services over to using bun instead of node, and would also like to use it as much as possible for a library of shared code for them. Do I use bun for everything but the `npm publish` command? Or is there a way to publish from bun?

**Reply by moshOntong-IT** [reply]
Do you found a solution?

**Comment by testerez** [comment]
May be a starting point: https://github.com/wobsoriano/bun-lib-starter (found here: https://github.com/oven-sh/awesome-bun#boilerplates)

**Reply by Alexnortung** [reply]
Thanks [bun-lib-starter](https://github.com/wobsoriano/bun-lib-starter) is a great starting point

**Comment by arshad-yaseen** [comment]
https://bunup.dev/

**Comment by ramdanegie** [comment]
Yes, you can absolutely use Bun to develop and publish npm packages. Here's the full workflow:

**1. Initialize**

```bash
bun init
```

This creates package.json, index.ts, and .gitignore.

**2. Structure**

```
my-package/
├── src/
│   └── index.ts
├── package.json
├── tsconfig.json
└── bun.lockb
```

**3. Build for npm**

```bash
bun build ./src/index.ts --outdir ./dist --target node
```

For TypeScript declarations, add a postbuild script:

```json
{
  "scripts": {
    "build": "bun build ./src/index.ts --outdir ./dist --target node",
    "postbuild": "tsc --emitDeclarationOnly --outDir dist"
  }
}
```

**4. package.json fields**

```json
{
  "name": "@yourscope/my-package",
  "version": "0.1.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": ["dist"],
  "scripts": {
    "build": "bun build ./src/index.ts --outdir ./dist --target node",
    "prepublishOnly": "bun run build"
  }
}
```

**5. About bun.lockb**

No action needed. npm automatically ignores lockfiles when packing the tarball.

**6. Publish**

```bash
bun publish  # Bun 1.1+
# or
bunx npm publish
```

Your users can install with npm install, yarn add, pnpm add, or bun add.

**7. Alternative**: bunup.dev wraps bun build with declarations, ESM/CJS dual output, and minification.

**Comment by hesprs** [comment]
By the way, how can I use bun to pack a package into a tarball like `npm pack`?

---

I know it now, it's `bun pm pack`.
